### PR TITLE
Install dependencies

### LINKED ISSUE #1215: [Question]: How to add app during deployment from the appDependencyProbingPaths
### Question

Hi,

My CICD builds,, but fail during deployment stage:

image

During the Build stage, the app got downloaded from the `appDependencyProbingPaths`
> Downloaded dependencies: D:\a\GTM-BC-9ARaptorDocumentManagement-W1\GTM-BC-9ARaptorDocumentManagement-W1\.dependencies\GTM-BC-9AAdvMan-License-main-Apps-1.1.14.0.zip, (D:\a\GTM-BC-9ARaptorDocumentManagement-W1\GTM-BC-9ARaptorDocumentManagement-W1\.dependencies\GTM-BC-9AAdvMan-License-main-TestApps-1.1.14.0.zip)
> Downloaded dependencies apps: D:\a\GTM-BC-9ARaptorDocumentManagement-W1\GTM-BC-9ARaptorDocumentManagement-W1\.dependencies\GTM-BC-9AAdvMan-License-main-Apps-1.1.14.0.zip
> Downloaded dependencies test apps: (D:\a\GTM-BC-9ARaptorDocumentManagement-W1\GTM-BC-9ARaptorDocumentManagement-W1\.dependencies\GTM-BC-9AAdvMan-License-main-TestApps-1.1.14.0.zip)

But during deployment, the action show this error:
> Deploying to UAT failed.  Status Code 422 : Unprocessable Entity Extension compilation failed error AL1024: A package with publisher '9altitudes', name '9A Advanced Manufacturing - License', and a version compatible with '1.0.0.0' could not be loaded

Is there another setting I forgot, so the apps in the `appDependencyProbingPaths` are deployed as well?

### PR DESCRIPTION
**Pull request by freddydk**
Fixes #1215 

Add new DeployTo setting called DependencyInstallMode, which can be ignore, install or upgrade. Default is install (which doesn't cause disruption with prior functionality)

If DependencyInstallMode is set to ignore - then dependencies will not be installed/upgraded.

All AppSource Apps will be installed/upgraded first, which can be done in any order as they are installed/upgraded with `-installOrUpdateNeededDependencies`. All dependency PTEs are installed afterwards in dependency order.

Rules for deploying dependencies to online environments goes here:

If the App (from dependencies) is a PTE:
- If it is installed in a newer version - leave it and see if it works
- If it is installed in an older version - Upgrade if DependencyInstallMode is upgrade (else issue WARNING)

If the App (from dependencies) is an AppSource App:
- If it isn't already installed - install it from AppSource
- If it is installed in an older version in the Dev scope - issue a warning and leave as is
- If it is installed in an older version - Upgrade if DependencyInstallMode is upgrade (else issue WARNING)
- If it is installed in a newer version - issue a warning and leave as is

TODO:
- [x] Add Release note
- [x] Add to settings.md

### CODE REVIEW

File: Actions/AL-Go-Helper.ps1 [hunk 1]
```diff
@@ -18,7 +18,7 @@ $defaultCICDPushBranches = @( 'main', 'release/*', 'feature/*' )
 [Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseDeclaredVarsMoreThanAssignments', 'defaultCICDPullRequestBranches', Justification = 'False positive.')]
 $defaultCICDPullRequestBranches = @( 'main' )
 $runningLocal = $local.IsPresent
-$defaultBcContainerHelperVersion = "preview" # Must be double quotes. Will be replaced by BcContainerHelperVersion if necessary in the deploy step - ex. "https://github.com/organization/navcontainerhelper/archive/refs/heads/branch.zip"
+$defaultBcContainerHelperVersion = "https://github.com/microsoft/navcontainerhelper/archive/refs/heads/freddydk/dependencies.zip" # Must be double quotes. Will be replaced by BcContainerHelperVersion if necessary in the deploy step - ex. "https://github.com/organization/navcontainerhelper/archive/refs/heads/branch.zip"
```
Comment (mazhelez) [inline Actions/AL-Go-Helper.ps1:21]: Don't forget to revert this.

Comment (freddydk) [reply Actions/AL-Go-Helper.ps1:21]: as soon as the containerhelper bug is merged and shipped

**Review by mazhelez — DISMISSED** [review]
LGTM

File: Tests/DetermineDeploymentEnvironments.Test.ps1 [hunk 2]
```diff
@@ -60,13 +60,13 @@ Describe "DetermineDeploymentEnvironments Action Test" {
...
+        $DeploymentEnvironmentsJson | ConvertFrom-Json | ConvertTo-HashTable -recurse | Should -MatchHashtable @{"test"=@{"EnvironmentType"="SaaS";"EnvironmentName"="test";"Branches"=@();"BranchesFromPolicy"=@();"Projects"="*";"DependencyInstallMode"="install";"Scope"=$null;"syncMode"=$null;"buildMode"=$null;"continuousDeployment"=$null;"runs-on"=@("ubuntu-latest");"shell"="pwsh";"ppEnvironmentUrl"="";"companyId"=""};"another"=@{"EnvironmentType"="SaaS";"EnvironmentName"="another";"Branches"=@();"BranchesFromPolicy"=@();"Projects"="*";"DependencyInstallMode"="install";"Scope"=$null;"syncMode"=$null;"buildMode"=$null;"continuousDeployment"=$null;"runs-on"=@("ubuntu-latest");"shell"="pwsh";"ppEnvironmentUrl"="";"companyId"=""}}
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Tests/DetermineDeploymentEnvironments.Test.ps1:63]: This is the test - tests that the default should match this hashtable.

File: Actions/Deploy/Deploy.ps1 [hunk 3]
```diff
@@ -10,6 +10,80 @@ Param(
...
+                    if ($installMode -eq 'upgrade') {
+                        Write-Host "$msg Needs upgrade."
+                        $needsUpgrade = $true
+                    }
+                    else {
+                        Write-Host "::WARNING::$msg Set DependencyInstallMode to 'upgrade' to upgrade dependencies."
+                    }
+                }
+                elseif ($newVersion -lt $installedVersion) {
+                    Write-Host "::WARNING::App $($appJson.name) is already installed in version $installedVersion, which is higher than $newVersion, used for this build."
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/Deploy/Deploy.ps1:50]: Have added a mitigation to the warning

**Review by aholstrup1 — DISMISSED** [review]

**Review by louagej — APPROVED** [review]

File: Actions/DetermineDeploymentEnvironments/DetermineDeploymentEnvironments.ps1 [hunk 4]
```diff
@@ -144,6 +144,7 @@ else {
             "Branches" = @()
             "BranchesFromPolicy" = @()
             "Projects" = '*'
+            "DependencyInstallMode" = "install"  # ignore, install or upgrade
```
Comment (aholstrup1) [inline Actions/DetermineDeploymentEnvironments/DetermineDeploymentEnvironments.ps1:147]: Shouldn't this (and the other settings) just be part of https://github.com/microsoft/AL-Go/blob/e5f2df3b796102c8ef4b018e797550e974b5554a/Actions/AL-Go-Helper.ps1#L552?

File: Actions/Deploy/Deploy.ps1 [hunk 5]
```diff
@@ -120,43 +213,48 @@ else {
...
-            if (!$sandboxEnvironment) {
-                throw "Scope Dev is only valid for sandbox environments"
-            }
-            $parameters = @{
-                "bcAuthContext" = $bcAuthContext
-                "environment" = $deploymentSettings.EnvironmentName
-                "appFile" = $apps
+        else {
+            if ($dependencies) {
+                InstallOrUpgradeApps -bcAuthContext $bcAuthContext -environment $deploymentSettings.EnvironmentName -Apps $dependencies
```
Comment (aholstrup1) [inline Actions/Deploy/Deploy.ps1:218]: We're not passing anything to the installMode parameter

File: Actions/Deploy/Deploy.ps1 [hunk 6]
```diff
@@ -10,6 +10,80 @@ Param(
...
+                else {
+                    Install-BcAppFromAppSource -bcAuthContext $bcAuthContext -environment $environment -appId $appJson.id -acceptIsvEula -installOrUpdateNeededDependencies
+                    # Update installed apps list as dependencies may have changed / been installed
+                    $installedApps = Get-BcInstalledExtensions -bcAuthContext $bcAuthContext -environment $environment | Where-Object { $_.isInstalled }
+                }
+            }
+        }
+        if ($PTEsToInstall) {
+            # Install or upgrade PTEs
+            Publish-PerTenantExtensionApps -bcAuthContext $bcAuthContext -environment $environment -appFiles $PTEsToInstall -SchemaSyncMode "Add"
```
Comment (aholstrup1) [inline Actions/Deploy/Deploy.ps1:79]: Should we use the syncMode parameter here instead of hardcoding it to "Add"?

Comment (freddydk) [reply Actions/Deploy/Deploy.ps1:79]: IMO dependencies should never be force updated - if anything we should have a DependencyInstallMode called UpgradeForce or like

File: Actions/Deploy/Deploy.ps1 (hunk 5 above)
Comment (freddydk) [reply Actions/Deploy/Deploy.ps1:218]: nice catch

File: Actions/DetermineDeploymentEnvironments/DetermineDeploymentEnvironments.ps1 (hunk 4 above)
Comment (freddydk) [reply Actions/DetermineDeploymentEnvironments/DetermineDeploymentEnvironments.ps1:147]: Problem here is that `DeployTo<environmentName>` cannot be pre-set in settings as we don't know the environment name.

This maybe could have been made differently (and more json strict) by using

```json
DeployTo: [
    {
        "environmentName": "environmentName",
        "settings": {
            "settings": xxx
        }
    }
]
```

And maybe we should shift to do something like that - but that will require us to have a mechanism to upgrade settings as part of upgrade AL-Go System Files.

That is not for this PR.

File: Actions/Deploy/Deploy.ps1 (hunk 6 above)
Comment (freddydk) [reply Actions/Deploy/Deploy.ps1:79]: will add this

File: Actions/Deploy/Deploy.ps1 [hunk 7]
```diff
@@ -10,6 +10,85 @@ Param(
...
+function InstallOrUpgradeApps {
+    Param(
+        [hashtable] $bcAuthContext,
+        [string] $environment,
+        [string[]] $apps,
+        [string] $installMode
+    )
+
+    $schemaSyncMode = "Add"
+    if ($installMode -eq 'ForceUpgrade') {
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/Deploy/Deploy.ps1:22]: sure

File: Actions/Deploy/Deploy.ps1 [hunk 8]
```diff
@@ -10,6 +10,85 @@ Param(
...
+        # Run through all apps and install or upgrade AppSource apps first (and collect PTEs)
+        foreach($app in $apps) {
+            # Get AppJson (works for full .app files, symbol files and also runtime packages)
+            $appJson = Get-AppJsonFromAppFile -appFile $app
+            $isPTE = ($appjson.idRanges.from -lt 100000 -and $appjson.idRanges.from -ge 50000)
+            $installedApp = $installedApps | Where-Object { $_.id -eq $appJson.id }
+            $needsUpgrade = $false
+            $needsInstall = $false
+            if ($installedApp) {
+                $newVersion = [version]::new($appJson.Version)
```
Comment (tomasmsft) [inline Actions/Deploy/Deploy.ps1:42]: consider extracting logic to local func CheckAppNeedsUpgrade

**Review by tomasmsft — APPROVED** [review]

File: Actions/Deploy/Deploy.ps1 [hunk 9]
```diff
@@ -10,6 +10,85 @@ Param(
...
+                $installedVersion = [version]::new($installedApp.versionMajor, $installedApp.versionMinor, $installedApp.versionBuild, $installedApp.versionRevision)
+                if ($newVersion -gt $installedVersion) {
+                    $msg = "Dependency app $($appJson.name) is already installed in version $installedVersion, which is lower than $newVersion."
+                    if ($installMode -eq 'upgrade') {
+                        Write-Host "$msg Needs upgrade."
+                        $needsUpgrade = $true
+                    }
+                    else {
+                        Write-Host "::WARNING::$msg Set DependencyInstallMode to 'upgrade' or 'forceUpgrade' to upgrade dependencies."
```
Comment (mazhelez) [inline Actions/Deploy/Deploy.ps1:51]: This warning might not be necessary. Even in _BCApps_ the dependencies are always "major.minor.0.0". 
I imagine bumping the dependency version is only needed if you want to uptake functionality from a newer version. 
In that sense, a dependency version is the "minimum required version", so doesn't need to be updated all the time.

Comment (freddydk) [reply Actions/Deploy/Deploy.ps1:51]: I will remove the github warning, but keep the output

**Review by tomasmsft — APPROVED** [review]

File: Actions/Deploy/Deploy.ps1 [hunk 10]
```diff
@@ -10,6 +10,96 @@ Param(
...
+            if ($installMode -eq 'upgrade') {
+                Write-Host "$msg Needs upgrade."
+                $needsUpgrade = $true
+            }
+            else {
+                Write-Host "::WARNING::$msg Set DependencyInstallMode to 'upgrade' or 'forceUpgrade' to upgrade dependencies."
+            }
+        }
+        elseif ($newVersion -lt $installedVersion) {
+            Write-Host "::WARNING::Dependency app $($appJson.name) is already installed in version $installedVersion, which is higher than $newVersion, used for this build. Please update your local copy of this dependency."
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/Deploy/Deploy.ps1:36]: Your comment was on the other warning, which I changed.

This warning states that they should update their local copy of the app with the updated app - in order to not run into problems later down the road.

**Review by mazhelez — APPROVED** [review]

**Pull request merged**